lib/events.js

```diff
@@ -497,29 +497,49 @@ function unwrapListeners(arr) {
 
 function once(emitter, name) {
   return new Promise((resolve, reject) => {
-    const eventListener = (...args) => {
-      if (errorListener !== undefined) {
-        emitter.removeListener('error', errorListener);
-      }
-      resolve(args);
-    };
-    let errorListener;
-
-    // Adding an error listener is not optional because
-    // if an error is thrown on an event emitter we cannot
-    // guarantee that the actual event we are waiting will
-    // be fired. The result could be a silent way to create
-    // memory or file descriptor leaks, which is something
-    // we should avoid.
-    if (name !== 'error') {
-      errorListener = (err) => {
-        emitter.removeListener(name, eventListener);
-        reject(err);
+    if (typeof emitter.addEventListener === 'undefined') {
+      const eventListener = (...args) => {
+        if (errorListener !== undefined) {
+          emitter.removeListener('error', errorListener);
+        }
+        resolve(args);
       };
+      let errorListener;
+
+      // Adding an error listener is not optional because
+      // if an error is thrown on an event emitter we cannot
+      // guarantee that the actual event we are waiting will
+      // be fired. The result could be a silent way to create
+      // memory or file descriptor leaks, which is something
+      // we should avoid.
+      if (name !== 'error') {
+        errorListener = (err) => {
+          emitter.removeListener(name, eventListener);
+          reject(err);
+        };
+
+        emitter.once('error', errorListener);
+      }
+
+      emitter.once(name, eventListener);
+      return;
+    }
+
+    if (typeof emitter.addEventListener === 'function') {
+      // EventTarget does not have `error` event semantics like Node
+      // EventEmitters, we do not listen to `error` events here.
+      if (name === 'error') {
+        reject('EventTarget does not have `error` event semantics');
+      }
```

Comment:
I would remove this and instead register a listener for the `'error'` event – on the web that does exist for some `EventTarget`s, and it matches what we do for `EventEmitter`.

(If you do have a synchronous rejection here, please use an Error object and either throw it or add a `return;` afterwards.)

Reply:
Yes, and in general we never reject with strings in core if we can help it :] 

There is a convention of how (and especially argument errors) work in node and our user base relies on it. In this case rejecting with a string means the receiver of the promise does not have a stack trace to figure out where the rejection came from.

Reply:
@addaleax will update the code to register for 'error' event and update the tests, seems reasonable to me.
@benjamingr thanks for the clarification.